### PR TITLE
(fleet) split the installer deamon from its package manager

### PR DESCRIPTION
**Pull request by arbll**
This PR cleanly splits the package manager from the daemon used for remote updates.

The daemon:
- Runs the long running service
- Handles RC
- Handles the local API

The package manager:
- Exposes a clean interface to manage packages
- Is isolated from the agent (go.mod in separate PR to reduce the number of reviewers)

Given how tangled everything was I unfortunately had to make a few code changes:
- Tests of the installer daemon were basically a 1-1 duplicate of the old `install.go` / package manager with RC on top. They now used a mock version of the package manager instead.
- I removed the global purge. It was dangerous in its current state 
- Lots of test re-wiring but coverage should be the exact same

A staging build of this branch is available and has gone through QA.

### CODE REVIEW

**Comment by arbll** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on

File: pkg/installer/packages/fixtures/server.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+		require.NoError(t, err)
+		err = remote.WriteIndex(ref, index)
+		require.NoError(t, err)
+
+		digest, err := index.Digest()
+		require.NoError(t, err)
+		f.indexDigest = digest.String()
+	}
+	return s
+}
```
Comment (BaptisteFoy) [inline pkg/installer/packages/fixtures/server.go:96]: **👏 Praise:** ‏Really nice

File: pkg/installer/packages/manager.go [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+	configDir := filepath.Join(m.configsDir, pkg.Name)
+	err = extractPackageLayers(pkg.Image, configDir, tmpDir)
+	if err != nil {
+		return fmt.Errorf("could not extract package layers: %w", err)
+	}
+	err = m.repositories.Create(ctx, pkg.Name, pkg.Version, tmpDir)
+	if err != nil {
+		return fmt.Errorf("could not create repository: %w", err)
+	}
+	return m.setupUnits(ctx, pkg.Name)
```
Comment (BaptisteFoy) [inline pkg/installer/packages/manager.go:153]: **🥜 Nitpick:** ‏It's not just units that are setup, we should probably rename to `setupPackage` or just `setup` for clarity

File: pkg/installer/packages/manager.go [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+func (m *managerImpl) PromoteExperiment(ctx context.Context, pkg string) error {
+	m.m.Lock()
+	defer m.m.Unlock()
+
+	repository := m.repositories.Get(pkg)
+	err := repository.PromoteExperiment(ctx)
+	if err != nil {
+		return fmt.Errorf("could not promote experiment: %w", err)
+	}
+	return m.stopExperiment(ctx, pkg)
```
Comment (BaptisteFoy) [inline pkg/installer/packages/manager.go:210]: **❓ Question:** (Not a blocker for this PR) Do you think it'd be worth handling the case where package-specific installation instructions differ from experiment failing & experiment succeeding?

For instance if package-written files (like /var/log or /etc files) would need to be moved or removed depending on the experiment success

File: pkg/installer/packages/manager.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+// Remove uninstalls a package.
+func (m *managerImpl) Remove(ctx context.Context, pkg string) error {
+	m.m.Lock()
+	defer m.m.Unlock()
+
+	err := m.repositories.Delete(ctx, pkg)
+	if err != nil {
+		return fmt.Errorf("could not remove package: %w", err)
+	}
+	return nil
```
Comment (BaptisteFoy) [inline pkg/installer/packages/manager.go:222]: **❓ Question:** Shouldn't we call a package-specific removal method here in `Remove`? For the injector removing the ld.preload.so config is essential. It should be done before deleting the package files on disk, fwiw.

File: pkg/installer/packages/manager.go [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+func (m *managerImpl) setupUnits(ctx context.Context, pkg string) error {
+	switch pkg {
+	case packageDatadogAgent:
+		return service.SetupAgentUnits(ctx)
+	case packageAPMInjector:
+		return service.SetupAPMInjector(ctx)
+	default:
+		return nil
+	}
+}
```
Comment (BaptisteFoy) [inline pkg/installer/packages/manager.go:264]: **💬 Suggestion:** ‏Let's the installer OCI setup too (let's do that in another PR)

File: pkg/installer/packages/service/datadog_agent.go [hunk 6]
```diff
@@ -83,10 +82,6 @@ func SetupAgentUnits(ctx context.Context) (err error) {
 			return
 		}
 	}
-	// write installinfo before start, or the agent could write it
-	if err = installinfo.WriteInstallInfo("updater_package", "manual_update"); err != nil {
-		return
-	}
```
Comment (BaptisteFoy) [inline pkg/installer/packages/service/datadog_agent.go:89]: **💬 Suggestion:** ‏Let's add it back

**Review by BaptisteFoy — APPROVED** [review]

File: pkg/installer/packages/manager.go (hunk 2 above)
Comment (arbll) [reply pkg/installer/packages/manager.go:153]: Let's do as a followup (probably by burning a lot of this entirely...)

Comment (arbll) [reply pkg/installer/packages/manager.go:153]: was previously the same, will address in a followup
